File: Algorithm/AlgorithmCSharp/AlgorithmCSharp/Q2.cs
using System;
using System.Linq;

namespace AlgorithmCSharp
{
    /// <summary>
    /// n クイーン問題
    /// </summary>
    class Q2
    {
        /// <summary>
        /// 全探索
        /// </summary>
        public class Zentansaku
        {
            private int count = 0;

            public void Start(int n)
            {
                // 生成した順列の配列の番号を行、値を列の番号とする
                var validPatternList = this.CreateJunretu(1, 9)
                    .Where(junretu => this.IsValid(junretu))
                    .ToArray();

                foreach (var validPattern in validPatternList)
                {
                    Console.WriteLine(string.Join(",", validPattern));
                }

                Console.WriteLine("パターン数:" + validPatternList.Length);
                Console.WriteLine("計算回数:" + this.count);
            }

            /// <summary>
            /// 指定した値の範囲内のすべてのパターンの順列を生成します。
            /// </summary>
            /// <param name="start"></param>
            /// <param name="last"></param>
            private int[][] CreateJunretu(int start, int last)
            {
                return this.CreateJunretu(Enumerable.Range(start, last - start + 1).ToArray(), new int[0]);
            }

            /// <summary>
            /// すべてのパターンの順列を生成します
            /// </summary>
            /// <param name="nokori"></param>
            /// <param name="junretu"></param>
            private int[][] CreateJunretu(int[] nokori, int[] junretu)
            {
                this.count++;
                if (!nokori.Any()) return new[] { junretu };

                return nokori.SelectMany(x =>
                {
                    var nn = nokori.Where(n => n != x).ToArray();
                    var jj = junretu.Concat(new[] { x }).ToArray();

                    return CreateJunretu(nn, jj);
                })
                .ToArray();
            }

            /// <summary>
            /// クイーンの位置が条件を満たしていればTrue、そうでなければFalse
            /// </summary>
            /// <param name="junretu"></param>
            /// <returns></returns>
            private bool IsValid(int[] junretu)
            {
                for (var i = 0; i < junretu.Length - 1; i++)
                {
                    for (var j = i + 1; j < junretu.Length; j++)
                    {
                        if (Math.Abs(i - j) == Math.Abs(junretu[i] - junretu[j])) return false;
                    }
                }

                return true;
            }
        }

        /// <summary>
        /// バックトラック
        /// </summary>
        public class BackTrack
        {
            private int count = 0;
            private int answer = 0;

            public void Start(int n)
            {
                // 生成した順列の配列の番号を行、値を列の番号とする
                this.CreateJunretu(1, n);
                Console.WriteLine("pata-nsuu:" + this.answer);
                Console.WriteLine("kaisuu:" + this.count);
            }

            /// <summary>
            /// 指定した値の範囲内のすべてのパターンの順列を生成します。
            /// </summary>
            /// <param name="start"></param>
            /// <param name="last"></param>
            private void CreateJunretu(int start, int last)
            {
                this.CreateJunretu(Enumerable.Range(start, last - start + 1).ToArray(), new int[0]);
            }

            /// <summary>
            /// すべてのパターンの順列を生成します
            /// </summary>
            /// <param name="nokori"></param>
            /// <param name="junretu"></param>
            private void CreateJunretu(int[] nokori, int[] junretu)
            {
                this.count++;
                if (junretu.Length >= 2)
                {
                    if (!this.IsValid(junretu)) return;

                    if (!nokori.Any())
                    {
                        this.answer++;
                        Console.WriteLine(string.Join(",", junretu));
                    }
                }

                foreach (var x in nokori)
                {
                    CreateJunretu(nokori.Where(n => n != x).ToArray(), junretu.Concat(new[] { x }).ToArray());
                }
            }

            /// <summary>
            /// クイーンの位置が条件を満たしていればTrue、そうでなければFalse
            /// </summary>
            /// <param name="junretu"></param>
            /// <returns></returns>
            private bool IsValid(int[] junretu)
            {
                for (var i = 0; i < junretu.Length - 1; i++)
                {
                    for (var j = i + 1; j < junretu.Length; j++)
                    {
                        if (Math.Abs(i - j) == Math.Abs(junretu[i] - junretu[j])) return false;
                    }
                }

                return true;
            }
        }
    }
}

File: Algorithm/AlgorithmFSharp/AlgorithmFSharp/Q2.fs
// n クイーン問題
module Q2
// 全探索
module Zentansaku =
    open System

    let rec createJunretuRec (nokori: int list) (junretu: int list) = 
        if (List.isEmpty nokori) then seq{ yield List.rev junretu }
        else
            seq{
                for x in nokori do
                    let nn = nokori |> List.filter(fun n -> not (n = x))
                    let jj = x :: junretu
                    yield! createJunretuRec nn jj }

    // 指定した範囲内のすべてのパターンの順列を生成します。
    let createJunretu start last = createJunretuRec [start..last] []

    let isValid (junretu : int list) =
        let lastIndex = (junretu |> List.length) - 1

        [0 .. lastIndex - 1] |> List.forall(fun i ->
            not ([i + 1 .. lastIndex] |> List.exists(fun j ->
                Math.Abs(i - j) = Math.Abs(junretu.[i] - junretu.[j]))))

    let start n = 
        printfn "全探索"
        let patternCount =
            createJunretu 1 n
            |> Seq.filter isValid
            |> Seq.map(fun x -> printfn "%A" x)
            |> Seq.length

        printfn "%dパターン" patternCount
            
// バックトラック
module BackTrack =
    open System
    
    let rec createJunretuRec (nokori: int list) (junretu: int list) isValid = 
        if (not (isValid junretu)) then seq{ yield [] }
        elif (List.isEmpty nokori) then seq{ yield List.rev junretu }
        else
            seq{
                for x in nokori do
                    let nn = nokori |> List.filter(fun n -> not (n = x))
                    let jj = x :: junretu
                    yield! createJunretuRec nn jj isValid }
            |> Seq.filter(fun x -> not (List.isEmpty x))

    // 指定した範囲内のすべてのパターンの順列を生成します。
    let createJunretu start last isValid = createJunretuRec [start..last] [] isValid

    let isValid (junretu : int list) =
        let lastIndex = (junretu |> List.length) - 1

        [0 .. lastIndex - 1] |> List.forall(fun i ->
            not ([i + 1 .. lastIndex] |> List.exists(fun j ->
                Math.Abs(i - j) = Math.Abs(junretu.[i] - junretu.[j]))))

    let start n = 
        printfn "バックトラック"
        let patternCount =
            createJunretu 1 n isValid
            |> Seq.filter isValid
            |> Seq.map(fun x -> printfn "%A" x)
            |> Seq.length

        printfn "%dパターン" patternCount

            

